Overhaul of div-alg code

olexandr-konovalov opened the pull request:
@drallenherman and @angeldelriomateos are now working on modernising the div-alg code. We are consolidating all work in this branch, and will from now on work on submitting small individual pull requests to this branch. This will give us opportunity to better review and test that changes work, and let the code stabilise here before we merge it into master and publish a new release. Should be also easier to resolve conflicts.

I have just submitted the first of the PRs to this branch in #59, and currently investigating why tests of manual examples fail in some older versions of GAP. I used that opportunity to test and document the contributing process - please see https://github.com/gap-packages/wedderga/wiki for instructions.

What would be useful to do before merging this:
- [ ] add more tests to improve code coverage
- [ ] remove commented out code which is no longer needed (if you need to preserve commented out code, I've rely on version control to keep its previous version; commenting out a block of code and writing a new code below makes diffs unclear, whereas editing IN the code makes changes more obvious)
- [ ] improve code formatting: not all code is properly indented. This should be done at the end, in a separate commit, not to be mixed with actual changes
- [ ] clean up use of GAPDoc elements in the documentation (`<A>`, `<C>` and `<M>`) instead of `<C>` in most places).

Commit:
Initial import of div-alg changes by Allen and Ángel

drallenherman committed:
Update version number in PackageInfo.g

We will temporarily use this version to distinguish the package
loaded from this branch

drallenherman committed:
Document Global Splitting and Character Descent Functions

Also extend LocalIndicesOfCyclotomicAlgebra mansection
and save the file without trailing whitespaces.

drallenherman committed:
Fix GaloisRepsOfCharacters; Improve GlobalCharacterDescent; Document new commands

angeldelriomateos committed:
New functions for cyclotomic algebras

IsCyclotomicAlgebra(K,F) decides whether K is a cyclotomic extension of
a NF.

ReducingCyclotomicAlgebra(A) = Reduces the crossed product of a cyc. alg.
as much as possible.

FactoringCycAlg(A) outputs [A[1],B,C] where the crossed product of A is
the tensor product of B and C.

drallenherman committed:
Remove trailing whitespaces

This was done accidentally because of default Atom settings.

angeldelriomateos committed:
Fix a bug in ReducingCyclotomicExtension

This bug was causing freezing when GSOCA was applied to some components of
SG(160,137).

Commit:
Fix GAPDoc syntax in the documentation

Now the manual builds, and examples are extracted into test files.

olexandr-konovalov commented on lib/div-alg.gi:
```diff
@@ -1706,25 +2209,599 @@ local W,l,W1,A,i;
+  y := Difference(v,x);
+  F1:=NF(con,List(y,j->A[4][j][2]));
+  F2:=NF(con,List(x,j->A[4][j][2]));
+  if IsCyclotomicExtension(F2,F) and IsCyclotomicExtension(F2,F) then    
```
The first F2 should be F1.

Commit:
Modify the version for proper loading (must end with "dev")

A version number ending in dev is considered to be infinite,
so "dev" should be the end.

angeldelriomateos commented:
I pushed the F2->F1 replacement. Why I can not see that commitment?

olexandr-konovalov commented:
@angeldelriomateos because you pushed it to some other place, which is no longer connected to this pull request. You need to update your setup now following https://github.com/gap-packages/wedderga/wiki - that's a more robust workflow, which would be good to adopt now. But if you wish, I can cherry-pick your commit from https://github.com/alex-konovalov/wedderga/commit/83cf230a851cd59022f343eb9503d0f040538f54 myself for now.

angeldelriomateos commented:
Ok. I am following the instructions cloning and everything else. So far I cloned the repository made the commitment and then try to push. The commit seems to be done but the push was rejected. I think that I have to continue with the instructions: fork,....

angeldelriomateos commented:
I understand that after cloning using    git clone https://github.com/gap-packages/wedderga.git I have to fork. For that, following the instruction I navigated to wedderga/divalg and clicked on fork. A message showed up that I alredy forked wedderga and mention angeldelriomateos/wedderga. 
I went there but the files there are old. 
What should I do then?

olexandr-konovalov commented:
@angeldelriomateos it does not matter that the fork is old - it matters that it exists so you can push branches to it. So go the next step, you're on the right way!

angeldelriomateos committed:
Fix the F2 to F1 replacement in ReducingCyclotomicAlgebra

angeldelriomateos commented:
It seems that I managed to do the F2->F1 change in divalg. I guess that Allen fixing is not commited yet because I cannot see it in the log.

olexandr-konovalov commented:
Thanks, you did that, indeed! Not completely following the pull request workflow, but for small obviously correct changes that's tolerable (although, one should never say "obviously correct" without testing - that's one of the nicest things for using pull requests; the other is ability to review the code by others BEFORE it is merged).

angeldelriomateos commented:
> Thanks, you did that, indeed! Not completely following the pull request workflow, but for small obviously correct changes that's tolerable (although, one should never say "obviously correct" without testing - that's one of the nicest things for using pull requests; the other is ability to review the code by others BEFORE it is merged).

I see. Instead of git commit -a I should have done git commit angeldelriomateos. Then I should follow the instructions for pull request. 
Learning one by one

olexandr-konovalov commented:
More precisely, you would have to do `git checkout -b fix-F2-F1`, then `git commit -a` and then `git push angeldelriomateos`. Getting there!

angeldelriomateos commented:
Still we are waiting for Allen push on fi.
After that there is an improvement that I think we should do, namely explode GSOCA to improve the output of WeddderburnDecompositionInfo. 
For example, at this moment the wedderburn decomposition of QS_3 looks like:
[ [ 1, Rationals ], [ 1, Rationals ], [ 1, Rationals, 3, [ 2, 2, 0 ] ] ]
With that improvement it would look as follow:
[ [ 1, Rationals ], [ 1, Rationals ], [ 2, Rationals ] ]
In theory also WedderburnDecomposition might be improved but this might require a harder work.

angeldelriomateos commented:
I found out where the fi bug is, so I am going to commit it myself.

angeldelriomateos committed:
Fix wrong placement of fi in GlobalSplittingOfCyclotomicAlgebra

angeldelriomateos commented:
> More precisely, you would have to do `git checkout -b fix-F2-F1`, then `git commit -a` and then `git push angeldelriomateos`. Getting there!

I followed this instructions and the last git push failed. I think that I didn't settle properly my username. 
Indeed, when I was doing git remote I wrote: 
git remote add angeldelriomateos https://github.com/your-username/wedderga.git
I think that I should have writen
git remote add angeldelriomateos https://github.com/angeldelriomateos/wedderga.git
If I do that now it says that angeldelriomateos have already exists
I guess that I invent a new nick name but this is probably a bad practice.
Can I go backward?

olexandr-konovalov commented:
@angeldelriomateos call this:
```
git remote set-url angeldelriomateos https://github.com/angeldelriomateos/wedderga.git
```
and then
```
git remote -v
```
to check that it had effect.

drallenherman commented:
> Still we are waiting for Allen push on fi.
> After that there is an improvement that I think we should do, namely explode GSOCA to improve the output of WeddderburnDecompositionInfo.
> For example, at this moment the wedderburn decomposition of QS_3 looks like:
> [ [ 1, Rationals ], [ 1, Rationals ], [ 1, Rationals, 3, [ 2, 2, 0 ] ] ]
> With that improvement it would look as follow:
> [ [ 1, Rationals ], [ 1, Rationals ], [ 2, Rationals ] ]
> In theory also WedderburnDecomposition might be improved but this might require a harder work.

It seems I missed a lot.  I had to go help Gwen make our Easter bread!  I might have it all to myself this year :-)  
Yes I was hoping we could apply GlobalSplitting to WDInfo so that only interesting simple components are produced.  But I think we need to discuss this with Sugandha and Gurmeet. 
It seems one test is still failing?  Alex said moving the fi: position might not be the only issue.  But that was back on #57.

olexandr-konovalov commented:
It seems @angeldelriomateos's #61 tests well - I will shortly merge it, and then rebase #59 on top of it - then we will see whether the `fi` problem will fix the tests...

angeldelriomateos committed:
Fix bug in ReducingCyclotomicAlgebra

The definition of the fields F1 and F2 have been modified.
F1=K^G1 and F2=K^G2 where K=F(E(m)) with F=A[2] and m=A[3].
The error consists in assuming that m coincides with the conductor of K.
G1 and G2 are expressed in terms of the information in A[4]
which in turn is expresed in terms of m. To define F1 and F2 properly
is necessarily to express the information of A[4] in terms of the conductor of K.
Another bug fixed in the return for the case where s1=s2=1, i.e. when the algebra is split.

drallenherman committed:
Adjust SchurIndexByCharacter; fix DefiningCharacterOfCyclotomicAlgebra

Also add documentation.

Closes #65.

drallenherman committed:
IsOrdinaryCharacter->IsCharacter and bugfix in SchurIndexByCharacter

angeldelriomateos committed:
Fix a bug in ReducingCyclotomicAlgebra: The order of the root of unity

This commit fixes a bug in the root of unity to be used for the cyclotomic
extension defining the coefficient field of one factor in the reducing
process of a cyclotomic algebra.

The function IsCyclotomicExtension has been modified so that the output is
a positive integer n if the input is of the form (K,F) with F an abelian
number field and K=F(E(n)). Otherwise the output is 0.

This is used in ReducingCyclotomicAlgebra for the production of the
factors of in the input as cyclotomic algebras. The factor Ai (i=1,2)
is a crossed product Fi*Gi, where Fi=F(E(ci)) and the info of the algebra
then takes the form [*,F,ci,*,*] where the last two entries are expressed
in terms of ci.

drallenherman committed:
Revise documentation and example for SchurIndexByCharacter

drallenherman committed:
Correct DefiningCharacterOfCyclotomicAlgebra to prevent test crashes

drallenherman committed:
Allow any number of generators in DefiningGroupAndCharacterOfCycAlg

This adjusts the character part of DefiningGroupAndCharacterOfCycAlg.

angeldelriomateos committed:
Rename IsCyclotomicExtension to CyclotomicExtensionGenerator

The old IsCyclotomicExtension was a boolean, the new one outputs
a positive integer n if the input is formed by K=F(E(n)) and an
abelian number field F.

An unrelated fix replaces E(con) by E(m), that was unnoticed in
the previous commit

Commit:
Start collecting further tests for div-alg code

angeldelriomateos committed:
Modify SimpleAlgebraInfoByData

This modification explodes ReducingCyclotomicAlgebra to shorten the output.
This has to be taken into account in the local index calculations,
cause otherwise it produces errors when asking for DefiningGroup.

Commit:
Add LocalIndicesOfCyclotomicAlgebra example from PR #64

drallenherman commented:
Testing some more ideas.  Most of the time needed for local indices of the harder groups order 1920 is  used by DefiningGroup at the g1:=IsomorphismSpecialPcGroup Step.  This takes almost 270 seconds, after which it takes 10 for the character table.  Replacing this with g1:=IsomorphismPermGroup it took 1.3 seconds but 13 to get the character table.  

We should change this to use PermGroup in DefiningGroup and DefiningGroupAndCharacter.

angeldelriomateos committed:
Fix a bug in LocalIndex calculation to adjust when DefiningGroup fails

angeldelriomateos committed:
Add SchurIndex of non factorizable algebra with four generators

angeldelriomateos commented:
> Testing some more ideas. Most of the time needed for local indices of the harder groups order 1920 is used by DefiningGroup at the g1:=IsomorphismSpecialPcGroup Step. This takes almost 270 seconds, after which it takes 10 for the character table. Replacing this with g1:=IsomorphismPermGroup it took 1.3 seconds but 13 to get the character table.
> 
> We should change this to use PermGroup in DefiningGroup and DefiningGroupAndCharacter.

I think that you should go ahead with this. This morning with the help of Alex I manage to update my version with Allen fixing in Galois descents and include my fixing on combination of SimpleAlgebraInfoWithData in some local index calculations and everything seems to be working smooth. The last improvement mentioned by Allen will be great if it shorten time calculation. I am very happy on how everything seems to be working now and this reduction in time calculation is welcomed.

Commit:
Adjust manual examples to new output

The output changed due to improvements made in PR #64.

angeldelriomateos committed:
Remove FactoringCycAlg

drallenherman committed:
Delete previously commented out lines of code

drallenherman committed:
Add documentation for CyclotomicExtensionGenerator (#75)

drallenherman committed:
Fix incorrect reduction mod m due to missing parentheses

angeldelriomateos committed:
Fix KillingCocycle, speed up ReducingCyclotomicAlgebra

Fix a bug in KillingCocycle to prevent a non-equivalent
algebra output. This bug showed up with SG(672,622).

Improve checking that a cyclic algebra (K/F,a) with a
root of unity is split by verifying that a is the norm
of a root of unity simply comparing the order of a with
the order of the norm of a generator of the group of
roots of unity of K.

drallenherman committed:
Update DecomposeCyclotomicAlgebra

drallenherman committed:
Improve KillingCocycle

drallenherman committed:
Fix a bug in DecomposeCyclotomicAlgebra

Before this, for certain inputs it could return cyclotomic algebras with zero factor set.

drallenherman committed:
Improve DecomposeCyclotomicAlgebra to not to fail as often

Commit:
Test for some special cases in DecomposeCyclotomicAlgebra

olexandr-konovalov commented:
@drallenherman @angeldelriomateos Hooray, with the new tests this PR now tests just above 90% of diffs in the code!

drallenherman committed:
Update DecomposeCyclotomicAlgebra test in divalg.tst

Commit:
Test length 4 and 5 cases in DecomposeCyclotomicAlgebra

drallenherman committed:
adjust LocalIndicesOfRationalSymbolAlgebra to avoid invalid input

It requires both inputs to be -1 or positive prime integers, and returns fail otherwise.

Commit:
Use elif in LocalIndicesOfRationalSymbolAlgebra

Also, fix indentation and split `if` statements into multiple lines.
Separate lines for different part of an `if` statement help to see
better which execution branches are not covered by tests.

olexandr-konovalov commented:
@drallenherman @angeldelriomateos is there anything left before we merge this and make a new release 4.10.0 or maybe even make it version 5.0.0 ?

angeldelriomateos commented:
I don't have anything new to add except maybe a library of decompositions. If Allen agrees one could merge everything and release a new version. Whether it should be 4.10.0 or 5.0.0 you know better what is the best alternative.

olexandr-konovalov commented:
@angeldelriomateos the library would be is a separate project, which should rather be done in a separate branch and submitted as a PR to master after merging `divalg`. Thus, merging this PR, making a new release, and then adding a library on top of the stable version of the package would be ideal.

drallenherman commented:
> I don't have anything new to add except maybe a library of decompositions. If Allen agrees one could merge everything and release a new version. Whether it should be 4.10.0 or 5.0.0 you know better what is the best alternative.

I don't have anything more to add.  I think we are ready to release this now.  I'm ok with calling it 4.10.0.  It doesn't make sense to call it 4.9.6 because we've added new features, but the main commands are still the same.

olexandr-konovalov commented:
Ok - then I will make a couple of edits in commit messages and then merge it.

olexandr-konovalov commented:
Merged! thanks @angeldelriomateos and @drallenherman !

olexandr-konovalov commented:
@drallenherman @angeldelriomateos version 4.10.0 now at https://gap-packages.github.io/wedderga/. It should be picked up by GAP package update system tomorrow for further testing.
